feat(ds-jobs): Toggleable confirmation dialog for 'Submit Job'

traeok opened the pull request:
Signed-off-by: Trae Yelovich <trae.yelovich@broadcom.com>

## Proposed changes

This is a feature implementation that follows up on issue #998. It adds a confirmation dialog for the "Submit Job" feature (right-click context menu for data set members), and is set to "All jobs" by default. This dialog can be configured from the extension settings - see the "Jobs: Confirm Submission" section.

| Proposed change | Screenshot |
| ------------------- | ------------ |
| Option in extension settings | ![image](https://user-images.githubusercontent.com/86500639/211363201-916ae57d-2e70-4015-8a55-d1062bb7e328.png) |
| Confirmation dialog | ![image](https://user-images.githubusercontent.com/86500639/211324106-bebecc6c-6c36-4b2d-ada9-9442cb6bdc08.png) |

## Release Notes

Milestone: 2.6.0

Changelog:
- Added confirmation message for "Submit Job" feature as an option in extension settings (set to "All jobs" by default) #998.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [x] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [x] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found

zFernand0 reviewed (commented):
LGTM! 😋 
Before approving...
Should we:
1. open a docs-site issue to include this new configuration in [Configuring Zowe Explorer](https://docs.zowe.org/stable/user-guide/ze-install#configuring-zowe-explorer)
2. use a notification pop-up vs modal (new-window)? (how does this affect Theia)?

zFernand0 commented on packages/zowe-explorer/package.json:
```diff
@@ -1710,6 +1710,12 @@
         },
+        "zowe.jobs.confirmSubmission": {
+          "type": "boolean",
+          "default": true,
```
🤔 Should the default be false? 
That way it's not a "breaking" change to the users

traeok replied:
Yes it should be `false`. Thanks for pointing that out, had it set to true for easy testing

zFernand0 reviewed (changes requested):
Will this be a breaking change?

traeok commented:
> LGTM! 😋 Before approving... Should we:
> 
> 1. open a docs-site issue to include this new configuration in [Configuring Zowe Explorer](https://docs.zowe.org/stable/user-guide/ze-install#configuring-zowe-explorer)?

Yep, works for me! 

> 2. use a notification pop-up vs modal (new-window)? (how does this affect Theia)?

Was debating this - not sure whether a modal will affect Theia behavior. Can do some testing to see if it will be affected, but always willing to switch to the cleaner notification-style dialogs.

traeok commented:
Looked into support for modal messages, and it seems that Theia will support the modal parameter:

```ts
/**
* Options to configure the message behavior.
*/
export interface MessageOptions {

  /*
   * Human-readable detail message that is rendered
   * less prominent. Note that detail is only shown
   * for modal messages.
   */
  detail?: string;

  /**
   * Indicates that this message should be modal.
   */
  modal?: boolean;
}
```

zFernand0 commented on packages/zowe-explorer/src/dataset/actions.ts:
```diff
@@ -974,10 +981,28 @@ export async function submitMember(node: api.IZoweTreeNode) {
+
+    if (confirmationOption !== "Disabled") {
+        const ownsJob = datasetName.split(".")[0] === nodeProfile.profile?.user?.toUpperCase();
+        if (confirmationOption === "All jobs") {
```
Is is possible to use constants for these options?
and should they be localized?

traeok replied:
Yep! Forgot about localization here 😅

zFernand0 commented on packages/zowe-explorer/src/dataset/actions.ts:
```diff
@@ -974,10 +981,28 @@ export async function submitMember(node: api.IZoweTreeNode) {
+                if (!(await showConfirmationDialog())) {
+                    return;
+                }
+            } else if (ownsJob && confirmationOption === "Your jobs") {
```
Not suggesting to remove this, but I'm wondering why someone will want only confirmation dialogs for their own jobs and not other user's jobs. 🤔

traeok replied:
One situation I considered is if a user is working on a critical mainframe job that is not ready to be submitted yet. In that case the user might want to protect themselves from an accidental submission of their own job.

zFernand0 commented on packages/zowe-explorer/package.json:
```diff
@@ -1711,8 +1711,20 @@
+            "All jobs"
+          ],
+          "enumDescriptions": [
+            "No confirmation dialogs will be shown when submitting a job.",
+            "Shows a confirmation dialog when submitting your jobs.",
+            "Shows a confirmation dialog when submitting jobs from another user.",
+            "Shows a confirmation dialog when submitting any job."
```
Localization question also applies to these descriptions 😋

traeok replied:
Thanks for pointing that out - going to look into localizing the strings mentioned.

zFernand0 reviewed (approved):
LGTM! 😋 
Left a couple of questions. 😅

JillieBeanSim commented:
@traeok could the code smells be addressed please.

zFernand0 commented:
Re: cognitive complexity
I finally understood why we have some `switch(true)`. All `case` statements are ignored towards the count.

General comment:
Sometimes trying to reduce cognitive complexity ends up decreasing the readability of a function 😅 

Happy to discuss during standup

JillieBeanSim commented:
> Re: cognitive complexity I finally understood why we have some `switch(true)`. All `case` statements are ignored towards the count.
> 
> General comment: Sometimes trying to reduce cognitive complexity ends up decreasing the readability of a function 😅
> 
> Happy to discuss during standup

I agree about the complexity reduction can lead to decreased readability sometimes but a lot of the time the complexity can be decreased easily by creating reusable methods or smaller private methods to handle a certain functional part of the code in the method flagged keeping it's readability.

JillieBeanSim reviewed (commented):
@traeok this works great and I like the drop-down settings.  should we also include a bit of info in the README under the [Submit Jobs section](https://github.com/zowe/vscode-extension-for-zowe/blob/main/packages/zowe-explorer/README.md#submit-a-jcl)

traeok commented:
> this works great and I like the drop-down settings. should we also include a bit of info in the README under the [Submit Jobs section](https://github.com/zowe/vscode-extension-for-zowe/blob/main/packages/zowe-explorer/README.md#submit-a-jcl)

@JillieBeanSim Thanks for the suggestion - I've updated the changelog to reflect the new drop-down options and added some documentation under that section. I can also replace the GIF w/ one showing the confirmation dialog, if desired.

JillieBeanSim commented on packages/zowe-explorer/CHANGELOG.md:
```diff
@@ -8,8 +8,8 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
-- Added confirmation message for "Submit Job" feature as an option in extension settings (disabled by default). [#998](https://github.com/zowe/vscode-extension-for-zowe/issues/998)
-- Added favorite job query to jobs view [#1947](https://github.com/zowe/vscode-extension-for-zowe/issues/1947)
+- Added favorite job query to jobs view. [#1947](https://github.com/zowe/vscode-extension-for-zowe/issues/1947)
+- Added confirmation message for "Submit Job" feature as an option in extension settings (set to "All jobs" by default). [#998](https://github.com/zowe/vscode-extension-for-zowe/issues/998)
```
thanks for catching the change needed for default value

JillieBeanSim reviewed (approved):
LGTM! thanks @traeok for the new feature
